float64 x
float64 y
float64 z
float32 delta_time